File: Rust/Copilot/NBodySimulation.rs
use std::f64::consts::PI;

const G: f64 = 6.67430e-11; // Gravitational constant
const DT: f64 = 1e-3; // Time step
const NUM_BODIES: usize = 1000000; // Number of bodies

#[derive(Clone, Copy)]
struct Body {
    x: f64,
    y: f64,
    z: f64,
    vx: f64,
    vy: f64,
    vz: f64,
    mass: f64,
}

fn initialize_bodies() -> Vec<Body> {
    let mut bodies = Vec::with_capacity(NUM_BODIES);
    bodies.push(Body {
        x: 0.0,
        y: 0.0,
        z: 0.0,
        vx: 0.0,
        vy: 0.0,
        vz: 0.0,
        mass: 1e30, // Central body mass
    });

    for i in 1..NUM_BODIES {
        let angle = 2.0 * PI * i as f64 / (NUM_BODIES - 1) as f64;
        bodies.push(Body {
            x: angle.cos() * 1e11,
            y: angle.sin() * 1e11,
            z: 0.0,
            vx: -angle.sin() * (G * bodies[0].mass / 1e11).sqrt(),
            vy: angle.cos() * (G * bodies[0].mass / 1e11).sqrt(),
            vz: 0.0,
            mass: 1e24, // Small body mass
        });
    }
    bodies
}

fn calculate_energy(bodies: &[Body]) -> f64 {
    let mut energy = 0.0;
    for body in bodies {
        let kinetic = 0.5 * body.mass * (body.vx.powi(2) + body.vy.powi(2) + body.vz.powi(2));
        let mut potential = 0.0;
        for other in bodies {
            if body != other {
                let dx = body.x - other.x;
                let dy = body.y - other.y;
                let dz = body.z - other.z;
                let distance = (dx.powi(2) + dy.powi(2) + dz.powi(2)).sqrt();
                potential -= G * body.mass * other.mass / distance;
            }
        }
        energy += kinetic + 0.5 * potential;
    }
    energy
}

fn kick_step(bodies: &mut [Body]) {
    for i in 0..bodies.len() {
        let mut ax = 0.0;
        let mut ay = 0.0;
        let mut az = 0.0;
        for j in 0..bodies.len() {
            if i != j {
                let dx = bodies[j].x - bodies[i].x;
                let dy = bodies[j].y - bodies[i].y;
                let dz = bodies[j].z - bodies[i].z;
                let distance = (dx.powi(2) + dy.powi(2) + dz.powi(2)).sqrt();
                let force = G * bodies[j].mass / distance.powi(3);
                ax += force * dx;
                ay += force * dy;
                az += force * dz;
            }
        }
        bodies[i].vx += ax * DT;
        bodies[i].vy += ay * DT;
        bodies[i].vz += az * DT;
    }

    for body in bodies.iter_mut() {
        body.x += body.vx * DT;
        body.y += body.vy * DT;
        body.z += body.vz * DT;
    }
}

fn main() {
    let mut bodies = initialize_bodies();

    let initial_energy = calculate_energy(&bodies);
    println!("Initial energy: {}", initial_energy);

    for _ in 0..1000 {
        kick_step(&mut bodies);
    }

    let final_energy = calculate_energy(&bodies);
    println!("Final energy: {}", final_energy);
}

File: Go/Copilot/NBodySimulation.go
package main

import (
	"fmt"
	"math"
)

const (
	G         = 6.67430e-11            // Gravitational constant
	DT        = 1e-3 * 3600 * 24 * 365 // Time step
	NumBodies = 10000                  // Number of bodies
)

type Body struct {
	x, y, z    float64
	vx, vy, vz float64
	mass       float64
}

func initializeBodies(bodies []Body) {
	bodies[0] = Body{0, 0, 0, 0, 0, 0, 1e30} // Central body mass

	for i := 1; i < len(bodies); i++ {
		angle := 2 * math.Pi * float64(i) / float64(len(bodies)-1)
		bodies[i] = Body{
			x:    math.Cos(angle) * 1e11,
			y:    math.Sin(angle) * 1e11,
			z:    0,
			vx:   -math.Sin(angle) * math.Sqrt(G*bodies[0].mass/1e11),
			vy:   math.Cos(angle) * math.Sqrt(G*bodies[0].mass/1e11),
			vz:   0,
			mass: 1e24 / float64(len(bodies)), // Small body mass
		}
	}
}

// func calculateEnergy(bodies []Body) float64 {
// 	energy := 0.0
// 	for i := range bodies {
// 		kinetic := 0.5 * bodies[i].mass * (bodies[i].vx*bodies[i].vx + bodies[i].vy*bodies[i].vy + bodies[i].vz*bodies[i].vz)
// 		potential := 0.0
// 		for j := range bodies {
// 			if i != j {
// 				dx := bodies[i].x - bodies[j].x
// 				dy := bodies[i].y - bodies[j].y
// 				dz := bodies[i].z - bodies[j].z
// 				distance := math.Sqrt(dx*dx + dy*dy + dz*dz)
// 				potential -= G * bodies[i].mass * bodies[j].mass / distance
// 			}
// 		}
// 		energy += kinetic + 0.5*potential
// 	}
// 	return energy
// }

func kickStep(bodies []Body) {
	for i := range bodies {
		ax, ay, az := 0.0, 0.0, 0.0
		for j := range bodies {
			if i != j {
				dx := bodies[j].x - bodies[i].x
				dy := bodies[j].y - bodies[i].y
				dz := bodies[j].z - bodies[i].z
				distance := math.Sqrt(dx*dx + dy*dy + dz*dz)
				force := G * bodies[j].mass / (distance * distance * distance)
				ax += force * dx
				ay += force * dy
				az += force * dz
			}
		}
		bodies[i].vx += ax * DT
		bodies[i].vy += ay * DT
		bodies[i].vz += az * DT
	}

	for i := range bodies {
		bodies[i].x += bodies[i].vx * DT
		bodies[i].y += bodies[i].vy * DT
		bodies[i].z += bodies[i].vz * DT
	}
}

func main() {
	bodies := make([]Body, NumBodies)
	initializeBodies(bodies)

	// initialEnergy := calculateEnergy(bodies)
	// fmt.Printf("Initial energy: %e\n", initialEnergy)

	for step := 0; step < 100; step++ {
		kickStep(bodies)
	}

	// finalEnergy := calculateEnergy(bodies)
	// fmt.Printf("Final energy: %e\n", finalEnergy)
	fmt.Printf("Body[0] %e %e %e", bodies[0].x, bodies[0].y, bodies[0].z)
}
